api/v1/flow/README.md

```diff
@@ -331,6 +331,7 @@ multiple fields are set, then all fields must match for the filter to match.
 | tcp_flags | [TCPFlags](#flow-TCPFlags) | repeated | tcp_flags filters flows based on TCP header flags |
 | node_name | [string](#string) | repeated | node_name is a list of patterns to filter on the node name, e.g. &#34;k8s*&#34;, &#34;test-cluster/*.domain.com&#34;, &#34;cluster-name/&#34; etc. |
 | ip_version | [IPVersion](#flow-IPVersion) | repeated | filter based on IP version (ipv4 or ipv6) |
+| trace_id | [string](#string) | repeated | trace_id filters flows by trace ID |
```

Comment:
I don't know much about the trace id semantics. The only available trace id is the one of the parent. Is it realistic to assume that get other trace IDs other than the parent? In which case, does this API still make sense if we have multiple trace IDs?

Reply:
This is a good question. The trace ID semantics we support are the ones defined by the [W3C Trace Context specification](https://www.w3.org/TR/trace-context/). The parent represents the incoming request in a tracing system and allows to identify the incoming request. The trace ID within the parent is an identifier that is used to uniquely identify a distributed trace. In contrast, the parent ID, which is more commonly known as the span ID, tracks a unit of work or operation. So a trace is made of one or more spans (or parents) that each contain a reference to the same trace ID to identify the trace.
In this model, I don't think there is room for multiple trace IDs (as in, spans/parents that belong to multiple traces). At least, this is my understanding of it. Let's maybe pull in someone more knowledgeable on the matter though :slightly_smiling_face: cc @chancez @lambdanis

Reply:
That's an interesting question. If context propagation is done only via `traceparent` header following the W3C spec - as @rolinh explained - each span has a single parent and all spans in a trace share the same trace ID. In other words, a trace is a tree of spans with the same trace ID.

That being said, trace context is not the only propagation method. There are a few "legacy" or vendor-specific ones, and there is a (fairly new) [W3C Baggage](https://www.w3.org/TR/baggage/) spec, supported by OpenTelemetry. Baggage can be used to pass links to multiple spans instead of just a parent (see e.g. [OpenTelemetry overview](https://opentelemetry.io/docs/reference/specification/overview/#links-between-spans)). A classic use case is batching, where multiple traces converge into a single operation.

So yeah, if we want Hubble to support different propagation methods than trace context, then it's possible to have multiple trace IDs. I'm not sure how widely baggage is used today, but I think it has a future, so I'd consider it realistic we eventually support it.

Regarding semantics of trace ID in flows - it's simply a trace ID that was propagated in the request. Currently we support only a single parent scenario. We may support many parents scenario (batching) or some other relationships, but I think flows filtering would be similar: check if the flow "belongs" to any of the traces defined by the trace IDs.

Reply:
Thanks for the detailed reply @lambdanis ! I think that the TLDR is that the current filtering implementation is probably good but we may need to revisit it down the road if/when add support for the baggage spec.